pulpcore/cli/container/repository.py

```diff
@@ -44,22 +51,62 @@
 
 def _tag_callback(ctx: click.Context, param: click.Parameter, value: str) -> str:
     if len(value) == 0:
-        raise click.ClickException("Please pass a non empty tag name.")
+        raise click.ClickException(_("Please pass a non empty tag name."))
     if re.match(VALID_TAG_REGEX, value) is None:
-        raise click.ClickException("Please pass a valid tag.")
+        raise click.ClickException(_("Please pass a valid tag."))
 
     return value
 
 
+def _source_callback(ctx: click.Context, param: click.Parameter, value: str) -> PulpEntityContext:
+    pulp_ctx = ctx.find_object(PulpContext)
+    assert pulp_ctx is not None
+    if len(value) == 0:
+        raise click.ClickException(_("Please pass in a non empty source repository."))
+
+    version: Optional[str] = None
+    pulp_href: Optional[str] = None
+    entity: Optional[EntityDefinition] = None
+    if value.startswith("/"):
+        pattern = rf"^{pulp_ctx.api_path}{PulpContainerRepositoryContext.HREF_PATTERN}"
+        if re.match(pattern, value) is None:
+            raise click.ClickException(
+                _("'{value}' is not a valid href for {option_name}.").format(
+                    value=value, option_name=param.name
+                )
+            )
+        pulp_href = value
+        if "versions" in value:
+            pulp_href = value.partition("versions")[0]
+            version = value.split("/")[-2]
+    else:
+        split_value = value.split(":", maxsplit=1)
+        if len(split_value) == 2:
+            version = split_value[1]
+        entity = {"name": split_value[0]}
+
+    repo_ctx = PulpContainerRepositoryContext(pulp_ctx, pulp_href=pulp_href, entity=entity)
+    if version:
+        # Check that the version makes sense
+        latest_version = int(repo_ctx.entity["latest_version_href"].split("/")[-2])
+        if not (0 < int(version) <= latest_version):
+            raise click.ClickException(
+                _(
+                    "Please specify a version that is less then the latest version {} and greater"
+                    " than 0"
+                ).format(latest_version)
+            )
+        repo_ctx.meta["version"] = version
```

Comment:
Can you just store what you want in meta? Because that's what I doing.

Reply:
Hmm good question, i don't even see it being used somewhere else.

Reply:
One concern i have with such things is the lifetime, once we get around to more pulp-cli-shell operations, where such a meta object could leak from one command into the next. But we are not there...

Reply:
I can pop the value from the meta object so that it goes back to its original state.

Reply:
Shouldn't we return a RepoVersionContext from this parameter?

Reply:
I did want to, but `meta`'s type definition is `Dict[str, str]`.

Reply:
Sorry, i meant the whole callback.

Reply:
Since the API takes either repository or repository version I decided attaching the version to a `RepositoryContext` was easiest way to determine which the user was specifying.

Reply:
Is specifying no version different from specifying the latest version?

Reply:
Looking at the serializer, no they are the same (except you can't specify a repository with 0 versions): https://github.com/pulp/pulp_container/blob/main/pulp_container/app/serializers.py#L527

Reply:
Updated to return `PulpContainerRepositoryVersionContext`.